### ISSUE TITLE
New segfault error in 0.9.4 (Feb 19, 2024)

### ISSUE THREAD

**Issue by Shawn-Tong** [3proxy/3proxy#1008]
I got this error message after updating from 0.9.4 (Jul 13, 2023) to 0.9.4 (Feb 19, 2024)
```
Feb 21 05:05:02 kernel: 3proxy[93978]: segfault at 7f8c00000000 ip 00007f8c00000000 sp 00007f8c0c227298 error 15
Feb 21 05:22:39 kernel: 3proxy[26344]: segfault at 7f0d00000000 ip 00007f0d00000000 sp 00007f0d9d0e6298 error 14
```

This segfault error happened in the system module, not in the 3proxy module. So I can't use `addr2line` to locate the crashing point like this: https://github.com/3proxy/3proxy/issues/976

Since I didn't use any plugins, I guess this bug should be related to these two commits:
1. Call local socket function: https://github.com/3proxy/3proxy/commit/375e3a74d0a06b8717115c77caae99ec865aaa8f
2. Adding state to socket functions: https://github.com/3proxy/3proxy/commit/8198db8617f9b3e01908302454649893c6ee70c4

Can you please check and fix it?
BTW, the #976 issue should be fixed. But what is the purpose of those two commits?

**Comment by Shawn-Tong** [comment]
Today I got a new segfault error
```
Feb 22 07:42:03 kernel: 3proxy[91229]: segfault at 7f2e14f844f0 ip 0000000000403d4e sp 00007f2e13c0fd90 error 4 in 3proxy[400000+2c000]
```

I used `addr2line` and found this crashing point in the function `freeparam()` of `proxymain.c`
```
	if(param->clisock != INVALID_SOCKET) {
		param->srv->so._shutdown(param->sostate, param->clisock, SHUT_RDWR); //// <---- crashed here ---- ////
		param->srv->so._closesocket(param->sostate, param->clisock);
	}
	myfree(param);
```

I hope it can help you to find the problem.

**Comment by z3APA3A** [comment]
Fixed with 35d1de6f5e141b33ab606b34e48dc688eacc1a8b

previous commits are required for robust TLS and another on-the-fly traffic processing functionality.
